cmd/node-joiner/nodejoiner_integration_test.go

```diff
@@ -0,0 +1,286 @@
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/rogpeppe/go-internal/testscript"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/yaml"
+
+	v1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/installer/internal/tshelpers"
+)
+
+const (
+	testResourcesFolder = "setup"
+)
+
+func TestMain(m *testing.M) {
+	// Set up the logger for testing
+	log.SetLogger(logr.Logger{})
+
+	os.Exit(testscript.RunMain(m, map[string]func() int{
+		"node-joiner": func() int {
+			if err := nodeJoiner(); err != nil {
+				return 1
+			}
+			return 0
+		},
+	}))
+}
+
+func TestNodeJoinerIntegration(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+	projectDir, err := os.Getwd()
+	assert.NoError(t, err)
+
+	testscript.Run(t, testscript.Params{
+		Dir: "testdata",
+
+		// Uncomment below line to help debug the testcases
+		// TestWork: true,
+
+		Deadline: time.Now().Add(10 * time.Minute),
+
+		Cmds: map[string]func(ts *testscript.TestScript, neg bool, args []string){
+			"isoCmp":              tshelpers.IsoCmp,
+			"isoIgnitionContains": tshelpers.IsoIgnitionContains,
+			"isoIgnitionUser":     tshelpers.IsoIgnitionUser,
+		},
+
+		Setup: func(e *testscript.Env) error {
+			// This is required for loading properly the embedded resources.
+			e.Cd = filepath.Join(projectDir, "../../data")
+
+			// Set the home dir within the test temporary working directory.
+			homeDir := filepath.Join(e.WorkDir, "home")
+			if err := os.Mkdir(homeDir, 0777); err != nil {
+				return err
+			}
+			for i, v := range e.Vars {
+				if v == "HOME=/no-home" {
+					e.Vars[i] = fmt.Sprintf("HOME=%s", homeDir)
+					break
+				}
+			}
+
+			// Reuse the current test tmp folder for envTest and fakeRegistry.
+			tmpDir := e.Getenv("TMPDIR")
+
+			// Create the fake registry
+			fakeRegistry := tshelpers.NewFakeOCPRegistry(tmpDir)
+
+			// Creates a new temporary cluster.
+			etcdDataDir, err := os.MkdirTemp(tmpDir, "etcd")
+			assert.NoError(t, err)
+			apiServerDataDir, err := os.MkdirTemp(tmpDir, "api-server")
+			assert.NoError(t, err)
+
+			testEnv := &envtest.Environment{
+				CRDDirectoryPaths: []string{
+					// Preload OpenShift specific CRDs.
+					filepath.Join(projectDir, "testdata", "setup", "crds"),
+				},
+				ErrorIfCRDPathMissing: true,
+
+				// Uncomment the following line if you wish to run the test without
+				// using the hack/go-integration-test-nodejoiner.sh script.
+				// BinaryAssetsDirectory: "/tmp/k8s/1.31.0-linux-amd64",
+
+				ControlPlane: envtest.ControlPlane{
```

Comment:
Is creating the temporary etcdDataDir and apiServerDataDir necessary? The docs say they are automatically created if unspecified. Perhaps this ControlPlane field can be removed.

Reply:
The idea was to keep all the folders within the testscript temp directory created for each test